types/carbon-components-react/lib/components/NumberInput/NumberInput.d.ts

```diff
@@ -20,7 +20,7 @@ export interface NumberInputProps extends Omit<ReactInputAttr, ExcludedAttribute
     label?: React.ReactNode,
     light?: boolean,
     size?: Extract<CarbonInputSize, "sm" | "xl">,
-    value: number,
+    value: number | '',
```

Comment:
I believe an empty string should only be allowed when allowEmpty is true, correct?

Reply:
Yes, empty value is only permitted when `allowEmpty` option is true. But, I don't know how to implement optional type when `allowEmpty` is true. Is my suggestion wrong?

Reply:
It's not wrong _per se_ but could be improved using Discriminated Union Type:

```ts
export interface NumberInputCommonProps
  extends Omit<ReactInputAttr, ExcludedAttributes>,
    InternationalProps<NumberInputTranslationKey> {
  allowEmpty?: boolean;
  ariaLabel?: string;
  helperText?: React.ReactNode;
  hideLabel?: boolean;
  iconDescription?: string;
  id: string;
  invalid?: boolean;
  invalidText?: React.ReactNode;
  isMobile?: boolean;
  label?: React.ReactNode;
  light?: boolean;
  size?: Extract<CarbonInputSize, 'sm' | 'xl'>;
  value: number;
  warn?: boolean;
  warnText?: React.ReactNode;
}

export type NumberInputProps = NumberInputCommonProps &
  ({ allowEmpty: true; value: number | '' } | { allowEmpty?: false; value: number });
```

See https://tsplay.dev/1WGg9W

Reply:
IMO I think this specific case is where the types can be loose and it will just boil down to proper usage/referencing the docs. Discriminated unions particularly with JSX can cause annoying issues. I think it's only worth introducing if the type signatures are significantly different (different props, different base element attributes, etc).

Reply:
I would also suggest the unioned types be their own exported interfaces as it will allow type augmentation and usage by the user.

Reply:
On second inspection it looks like more recent versions of TypeScript have improved inference. One example usage that seems to only fail on <= 3.7.x: [Playground](https://www.typescriptlang.org/play?ts=3.7.5#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQWAHpLHDAJ5hZwByAriABGWKAEkAdmD4wACsTDoAvHAAUAbxQAbTRADuAUXAcAXGyh8sAGjgA3ZJounxA4VDgAfOAHIvcAL4ecBr2OgZG7AD8pgT2qFa29o5wzkIi-gCU9LgQ4qjw-KliktKm2HgwAHQAYgDCADwFrhJSsvKoAHxwyirpXZ11ACbANgD07fR0wOIwIjG43ADy0iJyEApBdHBbWqGGYBxRcIIQEJo44vR+dHTZufAAsuw1JGClVJW1dUszUKsKnd1gto9HsOBk+httnBbnk4ABtFLWOIwAoAXS6mHeFT4cQAyjBkDMVAAGTKbbaUGB8KDiVTkqFbBouETNaQ7EHhRRqEIc-bsAJ2BxYLkpAJjelwdJ0Pw0IA).

Reply:
Then, you mean that @mmiszy 's suggestion should be accepted? or just continue with my suggestion? I'm not good at English, so I didn't understand your intention clearly. cc @kalbert312

Reply:
Continue with your suggestion.